File: quintus3.5/generic/qplib3.5/objects/demo/color.pl
%  SCCS   : @(#)color.pl	1.3 10/16/93
%  File   : color.pl
%  Author : Georges Saab
%  Purpose: foreign interface to color functions
%  Origin : 20 Apr 1993
%

:- module(color, [get_color/5]).

:- load_files(library(structs_decl), [when(compile_time),if(changed)]).

:- foreign_type
	widget        =  opaque.

foreign(get_color, c, get_color(+pointer(widget), +integer, +integer, 
	                         +integer, [-integer])).

foreign_file(system('color.o'), [get_color]).

:- load_foreign_files([system('color.o')], ['-lXm','-lXt','-lX11']),
   abolish(foreign_file, 2),
   abolish(foreign, 2).

File: quintus3.5/generic/qplib3.5/objects/demo/color.c
/*
  SCCS   : @(#)color.c	1.4 10/16/93
  File   : color.c
  Authors: Georges Saab
  Purpose: C definitions for allocating colormap entries
  Origin : 20 Apr 1993

*/

#include <X11/Intrinsic.h>
#include <X11/Xlib.h>
#include <string.h>

/*
	get_color() returns the color map entry for the color combination
specified by red, green, and blue for the display for widget widget.
Scaling is done so that this can be called multiple times with any 
combination of red green and blue in the range 0-100 without filling the
colormap entries for the display.
*/

int get_color(widget,red,green,blue)
    Widget *widget;
    int red, green, blue;
{
  XColor color;
  Display *display = XtDisplay(widget);
  int scr = DefaultScreen(display);
  Colormap cmap = DefaultColormap(display, scr);

  color.red   = ((int) red/17  )*13107;
  color.green = ((int) green/17)*13107;
  color.blue  = ((int) blue/17 )*13107;

  if (XAllocColor(display, cmap, &color))
      return (color.pixel);
  else {
      printf("Warning: Couldn't allocate requested color\n");
      return (BlackPixel(display, scr));
    }
}

